src/coreclr/src/vm/jithelpers.cpp

```diff
@@ -3001,84 +3003,66 @@ HCIMPL3(VOID, JIT_Unbox_Nullable, void * destPtr, CORINFO_CLASS_HANDLE type, Obj
+    {
+        // we allow enums and their primitive type to be interchangable
+        result = objRef->GetData();
+    }
+    else if (pMT1->IsEquivalentTo(pMT2))
     {
         // the structures are equivalent
         result = objRef->GetData();
     }
     else
     {
-        COMPlusThrowInvalidCastException(&objRef, TypeHandle(type));
+        COMPlusThrowInvalidCastException(&objRef, TypeHandle(pMT1));
     }
+
+    HELPER_METHOD_POLL(); 
     HELPER_METHOD_FRAME_END();
 
     return result;
 }
 HCIMPLEND
 
 /*************************************************************/
-/* the uncommon case for the helper below (allowing enums to be unboxed
-   as their underlying type */
-LPVOID __fastcall JIT_Unbox_Helper(CORINFO_CLASS_HANDLE type, Object* obj)
+/* Unbox helper that handles enums */
+HCIMPL2(LPVOID, Unbox_Helper, CORINFO_CLASS_HANDLE type, Object* obj)
 {
     FCALL_CONTRACT;
 
     TypeHandle typeHnd(type);
+    // boxable types have method tables
+    _ASSERTE(!typeHnd.IsTypeDesc());
 
-    CorElementType type1 = typeHnd.GetInternalCorElementType();
-
-        // we allow enums and their primtive type to be interchangable
+    MethodTable* pMT1 = typeHnd.AsMethodTable();
+    // must be a value type
+    _ASSERTE(pMT1->IsValueType());
 
     MethodTable* pMT2 = obj->GetMethodTable();
-    CorElementType type2 = pMT2->GetInternalCorElementType();
-    if (type1 == type2)
-    {
-        MethodTable* pMT1 = typeHnd.AsMethodTable();
-        if (pMT1 && (pMT1->IsEnum() || pMT1->IsTruePrimitive()) &&
-            (pMT2->IsEnum() || pMT2->IsTruePrimitive()))
-        {
-            _ASSERTE(CorTypeInfo::IsPrimitiveType_NoThrow(type1));
-            return(obj->GetData());
-        }
-    }
-
-    // Even less common cases (type equivalence) go to a framed helper.
-    ENDFORBIDGC();
-    return HCCALL2(JIT_Unbox_Helper_Framed, type, obj);
-}
-
-/*************************************************************/
-HCIMPL2(LPVOID, JIT_Unbox, CORINFO_CLASS_HANDLE type, Object* obj)
-{
-    FCALL_CONTRACT;
 
-    TypeHandle typeHnd(type);
-    VALIDATEOBJECT(obj);
-    _ASSERTE(!typeHnd.IsTypeDesc());   // boxable types have method tables
-
-        // This has been tuned so that branch predictions are good
-        // (fall through for forward branches) for the common case
-    if (obj != NULL) {
-        if (obj->GetMethodTable() == typeHnd.AsMethodTable())
-            return(obj->GetData());
-        else {
-                // Stuff the uncommon case into a helper so that
-                // its register needs don't cause spills that effect
-                // the common case above.
-            return JIT_Unbox_Helper(type, obj);
-        }
+    // we allow enums and their primitive type to be interchangable.
+    // if suspension is requested, defer to the framed helper.
+    CorElementType corElementType = pMT1->GetInternalCorElementType();
```

Comment:
```suggestion
    CorElementType corElementType = pMT1->GetVerifierCorElementType();
```
Multiple places. I had the same bug in my comment and fixed it, but I guess you have seen the original buggy version of the comment.

Reply:
It seemed that  `GetInternalCorElementType` actually correct here and cheaper than `GetVerifierCorElementType` 
- When we get here, since this is Unbox, the target type is a struct, enum or a true primitive
- We are not interested in structs, since they either should have matched already or need an equivalency check.
-  For others we just need the  `GetInternalCorElementType` to match.

From what I see the subtleness is about types that are primitives, but not true primitives. I am not sure what they are (byrefs?, void?), but I guess we should not see them here.
 

---
In reply to: [379897103](https://github.com/dotnet/runtime/pull/32353#discussion_r379897103) [](ancestors = 379897103)

Reply:
> types that are primitives, but not true primitives

These are structs that wrap a primitive type on x86.

In the early days of .NET Framework, we had a performance problem with structs that wrap a primitive type (there is a lot of struct like this). The JIT had close to none optimizations for structs at that time, and fixing it the right way was very hard. So we invented this quick hack: Treat them as the primitive type that they are wrapping. This is the type that `GetInternalCorElementType` returns. 

`GetInternalCorElementType` should be only ever used in places that deal with calling convention.

Reply:
If you seeing `GetVerifierCorElementType` to be more expensive than `GetInternalCorElementType `, you should be able to make `GetVerifierCorElementType` just return `GetInternalCorElementType` everywhere except x86. The two should return identical element type everywhere except x86.

Reply:
I see. Yes, `GetVerifierCorElementType` is noticeably more expensive.

Reply:
Hmmm, we are using the fake cor element types on !x86 as well in a few cases: https://github.com/dotnet/runtime/blob/4f9ae42d861fcb4be2fcd5d3d55d5f227d30e723/src/coreclr/src/vm/methodtablebuilder.cpp#L9759 . So using `GetInternalCorElementType` on !x86 may not be a good idea.

Reply:
I have tried 

```c++
    if (pMT1->GetInternalCorElementType() == pMT2->GetInternalCorElementType() &&
            (pMT1->IsEnum() || pMT1->IsTruePrimitive()) &&
            (pMT2->IsEnum() || pMT2->IsTruePrimitive()) &&
```
vs
```c++
    CorElementType corElementType = pMT1->GetVerifierCorElementType();
    if (corElementType != ELEMENT_TYPE_VALUETYPE &&  corElementType == pMT2->GetVerifierCorElementType() &&
```

The first variant is faster. (20%-30% on an Unbox  microbenchmark)
I think the checks are equivalent, unless I am missing something.